### PR TITLE
JobServerQ

### PR DESCRIPTION
**Pull request by TingluoHuang**

### CODE REVIEW

File: src/Agent.Worker/ExecutionContext.cs [hunk 1]
```diff
@@ -9,45 +9,221 @@
...
+            }
+
+            _record.CurrentOperation = currentOperation ?? _record.CurrentOperation;
+            _record.PercentComplete = Math.Max(percentage, _record.PercentComplete.Value);
+
+            _jobServerQueue.QueueTimelineRecordUpdate(_mainTimelineId, _record);
+        }
+
+        // This is not thread safe, the caller need to take lock before calling issue()
+        public void Issue(Issue issue)
```
Comment (bryanmacfarlane) [inline src/Agent.Worker/ExecutionContext.cs]: AddIssue

File: src/Agent.Worker/JobRunner.cs [hunk 2]
```diff
@@ -18,68 +18,91 @@ public sealed class JobRunner : AgentService, IJobRunner
...
-                .Where(x => string.Equals(x.HostType, hostType, StringComparison.OrdinalIgnoreCase))
-                .ToArray();
+            var jobServer = HostContext.GetService<IJobServer>();
+            await jobServer.ConnectAsync(ApiUtil.GetVssConnection(message));
 
-            // Add the prepare steps.
-            List<IStep> steps = new List<IStep>();
-            foreach (IJobExtension extension in extensions)
+            var jobServerQueue = HostContext.GetService<IJobServerQueue>();
+            jobServerQueue.Setup(message);
```
Comment (bryanmacfarlane) [inline src/Agent.Worker/JobRunner.cs]: This should be Start() - it's starting the pumping of the queues

File: src/Microsoft.VisualStudio.Services.Agent/JobServerQueue.cs [hunk 3]
```diff
@@ -1,59 +1,706 @@
...
+        // Job message information
+        private Guid _scopeIdentifier;
+        private string _hubName;
+        private Guid _planId;
+        private Guid _jobTimelineId;
+        private Guid _jobTimelineRecordId;
 
-        public void QueueWebConsoleLine(Guid timelineId, Guid timelineRecordId, string line)
+        // Default delay for Dequeue process
+        private static readonly TimeSpan _delayForWebConsoleLineDequeue = TimeSpan.FromMilliseconds(200);
```
Comment (bryanmacfarlane) [inline src/Microsoft.VisualStudio.Services.Agent/JobServerQueue.cs]: to top

File: src/Microsoft.VisualStudio.Services.Agent/JobServerQueue.cs [hunk 4]
```diff
@@ -1,59 +1,663 @@
...
+        private readonly object _fileUploadQueueLock = new object();
+        private readonly object _timelineUpdateQueueLock = new object();
+
+        // Task for each queue's dequeue process
+        private Task _webConsoleLineDequeueTask;
+        private Task _fileUploadDequeueTask;
+        private Task _timelineUpdateDequeueTask;
+
+        // Semaphore for each queue
+        private readonly SemaphoreSlim _webConsoleLineQueueSemaphore = new SemaphoreSlim(1, 1);
```
Comment (TingluoHuang) [inline src/Microsoft.VisualStudio.Services.Agent/JobServerQueue.cs]: 0,1

**Pull request merged**